sinon/lib/util/Wrapper.py

```diff
@@ -89,53 +88,136 @@ def wrapped(*args, **kwargs):
     return wrapped
 
-def __gen_index_list(condition, *args, **kwargs):
-    # Generating index list based on arguments and condition
-    # Note: ignore args[0] because it is callback in this condition
-    # combination
+def __gen_index_list(conditions, *args, **kwargs):
+    """
+    Args:
+        conditions: dictionary, the SinonStub's conditions (the user-defined behaviour for the stub)
+        args: tuple, the arguments inputed by the user
+        kwargs: tuple, the keyword arguments inputed by the user
+    Returns:
+        list, the list of indices in conditions for which the user args/kwargs match
+    """
+    return __get_call_indices(conditions["target"], args, kwargs, conditions["args"], conditions["kwargs"])
 
+def __get_call_indices(target, args, kwargs, args_list, kwargs_list):
+    """
+    Args:
+        target: object, the callback (self) argument for the owner of the stubbed function
+        args: tuple, the arguments inputed by the user
+        kwargs: tuple, the keyword arguments inputed by the user
+        args_list: list, a list of argument tuples
+        kwargs_list: list, a list of keyword argument tuples
+    Returns:
+        list, the list of indices in args_list/kwargs_list for which the user args/kwargs match
+    """
     # Todo: dirty hack
     if len(args) > 0:
-        if condition["target"] == args[0].__class__:
+        if target == args[0].__class__:
+            # Note: ignore args[0] because it is a callback (self) in this condition combination
             args = args[1:]
+            args_list = [i[1:] if i and i[0].__class__ == target else i for i in args_list]
 
     if args and kwargs:
-        if args in condition["args"] and kwargs in condition["kwargs"]:
-            args_indices = [i for i, x in enumerate(condition["args"]) if x == args]
-            kwargs_indices = [i for i, x in enumerate(condition["kwargs"]) if x == kwargs]
+        if args in args_list and kwargs in kwargs_list:
+            args_indices = [i for i, x in enumerate(args_list) if x == args]
+            kwargs_indices = [i for i, x in enumerate(kwargs_list) if x == kwargs]
             return list(set(args_indices).intersection(kwargs_indices))
     # args only
     elif args:
-        if args in condition["args"]:
-            return [i for i, x in enumerate(condition["args"]) if x == args and not condition["kwargs"][i]]
+        if args in args_list:
+            return [i for i, x in enumerate(args_list) if x == args and not kwargs_list[i]]
     #kwargs only
     elif kwargs:
-        if kwargs in condition["kwargs"]:
-            return [i for i, x in enumerate(condition["kwargs"]) if x == kwargs and not condition["args"][i]]
+        if kwargs in kwargs_list:
+            return [i for i, x in enumerate(kwargs_list) if x == kwargs and not args_list[i]]
+    return []
 
-def __gen_retfunc_with_args(index_list, condition, func, *args, **kwargs):
+def __get_call_count(target, args, kwargs, args_list, kwargs_list):
```

Comment:
As same as __gen_index_list, you can pass 3 parameters: args, kwargs, conditions for cleaner code

Reply:
For `__get_call_count`, the values for `args_list` and `kwargs_list` come from the `wrapper`. Would you prefer a signature like this?

```python
def __get_call_count(args, kwargs, conditions, wrapper):
```